exporter/tanzuobservabilityexporter/metrics_test.go

```diff
@@ -53,6 +54,60 @@ func TestMetricsConsumerNormal(t *testing.T) {
 	assert.Equal(t, 1, sender.numCloseCalls)
 }
 
+func TestMetricsConsumerNormalWithSourceTag(t *testing.T) {
+	sum := newMetric("sum", pdata.MetricDataTypeSum)
+	mockSumConsumer := &mockTypedMetricConsumer{typ: pdata.MetricDataTypeSum}
+	sender := &mockFlushCloser{}
+	tags := map[string]string{"source": "test_source", "test_key": "test_value"}
+	metrics := constructMetricsWithTags(tags, sum)
+	consumer := newMetricsConsumer(
+		[]typedMetricConsumer{mockSumConsumer}, sender, true)
+
+	assert.NoError(t, consumer.Consume(context.Background(), metrics))
+
+	assert.ElementsMatch(t, []string{"sum"}, mockSumConsumer.names)
+	assert.ElementsMatch(t, []string{"test_source"}, mockSumConsumer.sources)
+
+	assert.True(t, containsStr(mockSumConsumer.attrMaps, "test_key:test_value"))
+	assert.False(t, containsStr(mockSumConsumer.attrMaps, "source:test_source"))
+
+	assert.Equal(t, 1, mockSumConsumer.pushInternalMetricsCallCount)
+	assert.Equal(t, 1, sender.numFlushCalls)
+	assert.Equal(t, 0, sender.numCloseCalls)
+
+	consumer.Close()
+	assert.Equal(t, 1, sender.numCloseCalls)
+}
+
+func TestMetricsConsumerNormalWithHostnameTag(t *testing.T) {
+	sum := newMetric("sum", pdata.MetricDataTypeSum)
+	mockSumConsumer := &mockTypedMetricConsumer{typ: pdata.MetricDataTypeSum}
+	sender := &mockFlushCloser{}
+	tags := map[string]string{"host.name": "test_host.name", "hostname": "test_hostname"}
+	metrics := constructMetricsWithTags(tags, sum)
+	consumer := newMetricsConsumer(
+		[]typedMetricConsumer{mockSumConsumer}, sender, true)
+
+	assert.NoError(t, consumer.Consume(context.Background(), metrics))
+
+	assert.ElementsMatch(t, []string{"sum"}, mockSumConsumer.names)
+	assert.ElementsMatch(t, []string{"test_host.name"}, mockSumConsumer.sources)
+	assert.True(t, containsStr(mockSumConsumer.attrMaps, "hostname:test_hostname"))
```

Comment:
Looking at this a second time, I just realized that containsStr expects its first arg, the attrMaps, to be in sorted order.  I noticed that the attrMaps attribute is built by iterating over a map[string]interface{}. In go, maps are unordered, so when you iterate over a map, there is no guarantee that you will get the keys in sorted order.  While this test works now, it may break later on.

Reply:
resolved